libs/langgraph/langgraph/pregel/__init__.py

```diff
@@ -1339,79 +1356,109 @@ def update_state(
+                                local_read,
+                                step + 1,
+                                checkpoint,
+                                channels,
+                                managed,
+                                task,
+                                config,
+                            ),
+                        },
+                    ),
                 )
-                # if two nodes updated the state at the same time, it's ambiguous
-                if last_seen_by_node:
-                    if len(last_seen_by_node) == 1:
-                        as_node = last_seen_by_node[0][1]
-                    elif last_seen_by_node[-1][0] != last_seen_by_node[-2][0]:
-                        as_node = last_seen_by_node[-1][1]
-            if as_node is None:
-                raise InvalidUpdateError("Ambiguous update, specify as_node")
-            if as_node not in self.nodes:
-                raise InvalidUpdateError(f"Node {as_node} does not exist")
-            # create task to run all writers of the chosen node
-            writers = self.nodes[as_node].flat_writers
-            if not writers:
-                raise InvalidUpdateError(f"Node {as_node} has no writers")
-            writes: deque[tuple[str, Any]] = deque()
-            task = PregelTaskWrites((), as_node, writes, [INTERRUPT])
-            task_id = str(uuid5(UUID(checkpoint["id"]), INTERRUPT))
-            run = RunnableSequence(*writers) if len(writers) > 1 else writers[0]
-            # execute task
-            run.invoke(
-                values,
-                patch_config(
-                    config,
-                    run_name=self.name + "UpdateState",
-                    configurable={
-                        # deque.extend is thread-safe
-                        CONFIG_KEY_SEND: partial(
-                            local_write,
-                            writes.extend,
-                            self.nodes.keys(),
-                        ),
-                        CONFIG_KEY_READ: partial(
-                            local_read,
-                            step + 1,
-                            checkpoint,
-                            channels,
-                            managed,
-                            task,
-                            config,
-                        ),
-                    },
-                ),
-            )
+
             # save task writes
-            # channel writes are saved to current checkpoint
-            # push writes are saved to next checkpoint
-            channel_writes, push_writes = (
-                [w for w in task.writes if w[0] != PUSH],
-                [w for w in task.writes if w[0] == PUSH],
-            )
-            if saved and channel_writes:
-                checkpointer.put_writes(checkpoint_config, channel_writes, task_id)
+            for task in run_tasks:
+                channel_writes = [w for w in task.writes if w[0] != PUSH]
```

Comment:
can we collect all channel writes and push writes in a single loop here? and avoid another loop below?

Reply:
Unless I'm mistaken (cc @nfcampos) we need to put push writes after creating a new checkpoint? So we'd iterate anyway.